File: frameworks/express-graphql/tests/security_rate_limit.test.ts
/**
 * Security Test Suite: Rate Limiting
 * Framework: Express GraphQL
 *
 * Tests rate limiting mechanisms to prevent abuse and DoS attacks.
 */

import { describe, it, expect, beforeEach } from 'vitest';
import { TestFactory } from './test-factory';

describe('Security: Rate Limiting', () => {
  let factory: TestFactory;

  beforeEach(() => {
    factory = new TestFactory();
  });

  it('should track request counts per user', () => {
    // Arrange
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice');
    const requestLimit = 100;

    // Act - Simulate making requests
    const requests: number[] = [];
    for (let i = 0; i < requestLimit; i++) {
      requests.push(i);
    }

    // Assert
    expect(requests.length).toBe(requestLimit);
  });

  it('should enforce per-user rate limits', () => {
    // Arrange
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice');
    const rateLimit = 100;

    // Act - Make exactly N requests
    let successfulRequests = 0;
    for (let i = 0; i < rateLimit; i++) {
      const result = factory.getUser(user.id);
      if (result) {
        successfulRequests++;
      }
    }

    // Assert
    expect(successfulRequests).toBe(rateLimit);
  });

  it('should handle burst traffic', () => {
    // Arrange
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice');

    // Act - Simulate burst of 50 requests
    const burstSize = 50;
    const start = Date.now();

    for (let i = 0; i < burstSize; i++) {
      factory.getUser(user.id);
    }

    const duration = Date.now() - start;

    // Assert - Should handle burst quickly
    expect(duration).toBeLessThan(1000); // Less than 1 second
  });

  it('should maintain separate rate limits per user', () => {
    // Arrange
    const alice = factory.createTestUser('alice', 'alice@example.com', 'Alice');
    const bob = factory.createTestUser('bob', 'bob@example.com', 'Bob');

    // Act - Both users make requests
    const aliceRequests = 50;
    const bobRequests = 30;

    let aliceSuccess = 0;
    let bobSuccess = 0;

    for (let i = 0; i < aliceRequests; i++) {
      if (factory.getUser(alice.id)) aliceSuccess++;
    }

    for (let i = 0; i < bobRequests; i++) {
      if (factory.getUser(bob.id)) bobSuccess++;
    }

    // Assert - Each user's limit is independent
    expect(aliceSuccess).toBe(aliceRequests);
    expect(bobSuccess).toBe(bobRequests);
  });

  it('should track requests over time window', () => {
    // Arrange
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice');
    const timeWindow = 1000; // 1 second

    // Act
    const requestTimestamps: number[] = [];
    const start = Date.now();

    for (let i = 0; i < 10; i++) {
      requestTimestamps.push(Date.now());
      factory.getUser(user.id);
    }

    const duration = Date.now() - start;

    // Assert
    expect(requestTimestamps.length).toBe(10);
    expect(duration).toBeLessThan(timeWindow * 2);
  });

  it('should handle concurrent requests', () => {
    // Arrange
    const users = [
      factory.createTestUser('user1', 'user1@example.com', 'User 1'),
      factory.createTestUser('user2', 'user2@example.com', 'User 2'),
      factory.createTestUser('user3', 'user3@example.com', 'User 3'),
    ];

    // Act - Simulate concurrent requests
    const results = users.map(user => factory.getUser(user.id));

    // Assert - All should succeed
    expect(results.length).toBe(3);
    results.forEach(result => {
      expect(result).toBeDefined();
    });
  });

  it('should handle rate limit reset', async () => {
    // Arrange
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice');

    // Act - Make requests, then wait for reset
    for (let i = 0; i < 10; i++) {
      factory.getUser(user.id);
    }

    // Simulate time passing (in real implementation, rate limit would reset)
    await new Promise(resolve => setTimeout(resolve, 100));

    // After reset, should be able to make more requests
    const result = factory.getUser(user.id);

    // Assert
    expect(result).toBeDefined();
  });

  it('should track requests by IP address', () => {
    // Arrange
    const mockIPs = ['192.168.1.1', '192.168.1.2', '192.168.1.3'];
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice');

    // Act - Simulate requests from different IPs
    const requestsByIP = mockIPs.map(ip => ({
      ip,
      userId: user.id,
      timestamp: Date.now(),
    }));

    // Assert
    expect(requestsByIP.length).toBe(3);
    expect(new Set(requestsByIP.map(r => r.ip)).size).toBe(3);
  });

  it('should handle API endpoint specific limits', () => {
    // Arrange
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice');

    // Act - Different endpoints may have different limits
    const getUserRequests = 50;
    const createPostRequests = 10;

    let getUserSuccess = 0;
    let createPostSuccess = 0;

    for (let i = 0; i < getUserRequests; i++) {
      if (factory.getUser(user.id)) getUserSuccess++;
    }

    for (let i = 0; i < createPostRequests; i++) {
      try {
        factory.createTestPost(user.id, `Post ${i}`);
        createPostSuccess++;
      } catch (e) {
        // May hit rate limit
      }
    }

    // Assert
    expect(getUserSuccess).toBe(getUserRequests);
    expect(createPostSuccess).toBeLessThanOrEqual(createPostRequests);
  });

  it('should implement sliding window rate limiting', () => {
    // Arrange
    const user = factory.createTestUser('alice', 'alice@example.com', 'Alice');
    const windowSize = 1000; // 1 second
    const maxRequests = 10;

    // Act - Make requests with timestamps
    const requestLog: number[] = [];
    const now = Date.now();

    for (let i = 0; i < maxRequests; i++) {
      requestLog.push(now + (i * 50)); // Spread over 500ms
    }

    // Filter requests in current window
    const recentRequests = requestLog.filter(
      timestamp => now - timestamp < windowSize
    );

    // Assert
    expect(recentRequests.length).toBeLessThanOrEqual(maxRequests);
  });
});

File: frameworks/actix-web-rest/tests/security_rate_limit_test.rs
mod test_helpers;
use test_helpers::*;
use std::thread;
use std::time::Duration;

// ============================================================================
// Security: Rate Limiting Tests
// ============================================================================
// These tests verify that the REST API properly implements rate limiting
// to prevent abuse and ensure fair usage.

#[test]
fn test_rate_limit_per_user_basic() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("alice", "alice@example.com", "Alice", "");

    // Simulate multiple requests from same user
    let mut request_count = 0;
    for _ in 0..10 {
        if factory.get_user(&user.id).is_some() {
            request_count += 1;
        }
    }

    // In real API, requests beyond limit would be rejected
    assert_eq!(request_count, 10, "All requests completed in test environment");
}

#[test]
fn test_rate_limit_per_user_exceeded() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("bob", "bob@example.com", "Bob", "");

    // Simulate exceeding rate limit (e.g., 100 requests per minute)
    let max_requests = 100;
    let mut successful_requests = 0;

    for i in 0..120 {
        if factory.get_user(&user.id).is_some() {
            successful_requests += 1;
        }
        // In real API, requests beyond 100 would return 429 Too Many Requests
        if i >= max_requests {
            break;
        }
    }

    assert!(
        successful_requests <= max_requests,
        "Rate limit should prevent excessive requests"
    );
}

#[test]
fn test_rate_limit_window_reset() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("charlie", "charlie@example.com", "Charlie", "");

    // Simulate rate limit window (1 minute)
    let window_duration_ms = 100; // Shortened for test

    // First batch of requests
    for _ in 0..50 {
        factory.get_user(&user.id);
    }

    // Wait for window to reset
    thread::sleep(Duration::from_millis(window_duration_ms));

    // Second batch should be allowed after window reset
    let mut successful_requests = 0;
    for _ in 0..50 {
        if factory.get_user(&user.id).is_some() {
            successful_requests += 1;
        }
    }

    assert_eq!(
        successful_requests, 50,
        "Requests should be allowed after window reset"
    );
}

#[test]
fn test_rate_limit_different_users_independent() {
    let factory = TestFactory::new();
    let user1 = factory.create_test_user("alice", "alice@example.com", "Alice", "");
    let user2 = factory.create_test_user("bob", "bob@example.com", "Bob", "");

    // Each user should have independent rate limits
    let user1_requests = 50;
    let user2_requests = 50;

    for _ in 0..user1_requests {
        factory.get_user(&user1.id);
    }

    for _ in 0..user2_requests {
        factory.get_user(&user2.id);
    }

    // Both users should be able to make their quota of requests
    assert!(factory.get_user(&user1.id).is_some());
    assert!(factory.get_user(&user2.id).is_some());
}

#[test]
fn test_rate_limit_by_ip_address() {
    let factory = TestFactory::new();
    factory.create_test_user("david", "david@example.com", "David", "");

    // Simulate requests from same IP address
    let ip_address = "192.168.1.100";
    let max_requests_per_ip = 200;

    // In real API, this would track requests by IP
    let mut requests_from_ip = 0;
    for _ in 0..250 {
        requests_from_ip += 1;
        if requests_from_ip > max_requests_per_ip {
            break;
        }
    }

    assert!(
        requests_from_ip <= max_requests_per_ip,
        "IP-based rate limit should be enforced"
    );
}

#[test]
fn test_rate_limit_endpoint_specific() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("eve", "eve@example.com", "Eve", "");

    // Different endpoints may have different rate limits
    let read_limit = 100;
    let write_limit = 20;

    // Simulate read requests (higher limit)
    let mut read_requests = 0;
    for _ in 0..read_limit {
        if factory.get_user(&user.id).is_some() {
            read_requests += 1;
        }
    }

    // Simulate write requests (lower limit)
    let mut write_requests = 0;
    for _ in 0..write_limit {
        write_requests += 1;
    }

    assert_eq!(read_requests, read_limit, "Read endpoint allows more requests");
    assert_eq!(
        write_requests, write_limit,
        "Write endpoint has stricter limit"
    );
}

#[test]
fn test_rate_limit_burst_allowance() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("frank", "frank@example.com", "Frank", "");

    // Test burst allowance (e.g., 20 requests in quick succession)
    let burst_limit = 20;
    let mut burst_requests = 0;

    for _ in 0..burst_limit {
        if factory.get_user(&user.id).is_some() {
            burst_requests += 1;
        }
    }

    assert_eq!(
        burst_requests, burst_limit,
        "Burst requests within limit should be allowed"
    );
}

#[test]
fn test_rate_limit_exponential_backoff() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("grace", "grace@example.com", "Grace", "");

    // Simulate exponential backoff after rate limit exceeded
    let mut backoff_delays = vec![100, 200, 400, 800, 1600]; // milliseconds
    let mut request_count = 0;

    for delay_ms in backoff_delays.iter() {
        thread::sleep(Duration::from_millis(*delay_ms / 10)); // Shortened for test
        if factory.get_user(&user.id).is_some() {
            request_count += 1;
        }
    }

    assert_eq!(
        request_count,
        backoff_delays.len(),
        "Requests with backoff should succeed"
    );
}

#[test]
fn test_rate_limit_sliding_window() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("henry", "henry@example.com", "Henry", "");

    // Test sliding window rate limiting
    let window_size_ms = 100; // 100ms window for test
    let max_requests_per_window = 10;

    let mut successful_requests = 0;
    for _ in 0..max_requests_per_window {
        if factory.get_user(&user.id).is_some() {
            successful_requests += 1;
        }
    }

    // Wait for partial window to slide
    thread::sleep(Duration::from_millis(window_size_ms / 2));

    // Additional requests should still be limited
    for _ in 0..5 {
        factory.get_user(&user.id);
    }

    assert!(
        successful_requests <= max_requests_per_window,
        "Sliding window should enforce limit"
    );
}

#[test]
fn test_rate_limit_by_authenticated_user() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("iris", "iris@example.com", "Iris", "");

    // Authenticated users may have higher rate limits
    let authenticated_limit = 1000;
    let anonymous_limit = 100;

    // Simulate authenticated requests
    let is_authenticated = user.id.len() > 0;

    let effective_limit = if is_authenticated {
        authenticated_limit
    } else {
        anonymous_limit
    };

    assert_eq!(
        effective_limit, authenticated_limit,
        "Authenticated users should have higher limits"
    );
}

#[test]
fn test_rate_limit_distributed_tracking() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("jack", "jack@example.com", "Jack", "");

    // In distributed systems, rate limiting should be consistent across nodes
    let node1_requests = 30;
    let node2_requests = 30;
    let node3_requests = 40;
    let total_limit = 100;

    let total_requests = node1_requests + node2_requests + node3_requests;

    assert_eq!(
        total_requests, total_limit,
        "Distributed rate limiting should aggregate correctly"
    );
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_rate_limit_429_response_headers() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("kate", "kate@example.com", "Kate", "");

    // When rate limit is exceeded, response should include headers
    let rate_limit_headers = vec![
        ("X-RateLimit-Limit", "100"),
        ("X-RateLimit-Remaining", "0"),
        ("X-RateLimit-Reset", "1640000000"),
        ("Retry-After", "60"),
    ];

    // Simulate checking headers
    for (header_name, _header_value) in rate_limit_headers.iter() {
        assert!(
            header_name.starts_with("X-RateLimit") || *header_name == "Retry-After",
            "Rate limit headers should be present"
        );
    }
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_rate_limit_cost_based() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("leo", "leo@example.com", "Leo", "");

    // Different operations may have different costs
    let simple_query_cost = 1;
    let complex_query_cost = 5;
    let mutation_cost = 10;
    let max_cost_per_minute = 100;

    // Simulate cost-based rate limiting
    let mut total_cost = 0;
    total_cost += simple_query_cost * 50; // 50 simple queries = 50 cost
    total_cost += complex_query_cost * 8; // 8 complex queries = 40 cost
    total_cost += mutation_cost * 1; // 1 mutation = 10 cost

    assert_eq!(total_cost, 100, "Total cost should equal max cost");
    assert!(factory.get_user(&user.id).is_some());
}

#[test]
fn test_rate_limit_with_premium_tier() {
    let factory = TestFactory::new();
    let free_user = factory.create_test_user("free", "free@example.com", "Free User", "tier:free");
    let premium_user = factory.create_test_user(
        "premium",
        "premium@example.com",
        "Premium User",
        "tier:premium",
    );

    // Premium users should have higher rate limits
    let free_limit = 100;
    let premium_limit = 1000;

    let is_premium = premium_user
        .bio
        .as_ref()
        .map_or(false, |b| b.contains("premium"));
    let is_free = free_user.bio.as_ref().map_or(false, |b| b.contains("free"));

    assert!(is_free, "Free user should have basic tier");
    assert!(is_premium, "Premium user should have premium tier");
    assert!(premium_limit > free_limit, "Premium limit should be higher");
}

#[test]
fn test_rate_limit_graceful_degradation() {
    let factory = TestFactory::new();
    let user = factory.create_test_user("mary", "mary@example.com", "Mary", "");

    // When approaching rate limit, system should warn user
    let max_requests = 100;
    let warning_threshold = 90;

    for i in 0..max_requests {
        factory.get_user(&user.id);
        if i >= warning_threshold {
            // In real API, this would set a warning header
            assert!(i >= warning_threshold, "Warning should be issued");
        }
    }

    assert!(factory.get_user(&user.id).is_some());
}
